cuesubmit/cuesubmit/ui/Submit.py

```diff
@@ -103,7 +103,7 @@ def __init__(self, settingsWidgetType, jobTypes=JobTypes.JobTypes, parent=None,
             validators=[Validators.matchNoSpecialCharactersOnly, Validators.moreThan3Chars,
                         Validators.matchNoSpaces]
         )
-        shows = Util.getShows()
+        shows = Util.getShows() or ['no-show-exists'] # to limit out of range error if no show exists
```

Comment:
Thanks for sending this!

Couple thoughts:

- If you'd like, go ahead and update the error message to remove that SPI-specific stuff. I don't think contact info is very important here, I would just say something generic like:

  > The <show name> show does not exist. Please create it first or have your the administrator of your OpenCue deployment create it.

  That code is in [JobLauncher.java](https://github.com/AcademySoftwareFoundation/OpenCue/blob/e83c368929e1f07096798cea72c7891f506d1732/cuebot/src/main/java/com/imageworks/spcue/service/JobLauncher.java#L172), and there's a similar error message right above it too that could use updating.

- Rather than put arbitrary data into the array and allow the submission to fail on the Cuebot side (which IIUC requires the user to fill out all of the other submission fields first -- might be annoying), I think I'd prefer to see us check the value of `shows` and raise an error right here. Something like:

  ```
  shows = Util.getShows()
  if not shows:
    message = 'No shows exist yet. Please create one before attempting to submit a job.'
    # Haven't tested this, and I'm not sure if it's necessary to create this messageBox, or if raising an
    # exception will already cause that to happen.
    Widgets.messageBox(message, title='No Shows Exist', parent=self).show()
    raise opencue.exception.CueException(message)
  ```

Reply:
Thanks for the feedback :).

Removed hard coded value in java file.

I don't like that exception there too much as its in the middle of building of UI and if thrown, no messageBox gets created and it dies in similar fashion as previously.
I would have to check for this issue after UI is finished and throw exception then, but I don't think its totally necessary to have exception there. 

If I am missing something obvious, or preferred approach how to implement it, please let me know. I was looking around and there is only one exception thrown after submission error, which doesn't result in complete crash of UI, and it makes sense there.

Reply:
Good point -- I forgot this was happening during the initial app startup, yeah definitely throwing an exception there will cause problems :)  

A few thoughts:
- It would be great to show an exception somehow in the event the GUI is unavailable or not working as intended. My initial thought was: if this failure happens during unit tests (where we use a headless node with a virtual display) we won't be able to see the actual cause of failure.
- When I tried this locally, the dialog popped up way off to the side of my screen, which was confusing -- it took me a while to actually see it and in the meantime the app appeared to be locked up. I think this was due to creating the dialog before the app is fully shown.

Here's what I came up with: https://github.com/AcademySoftwareFoundation/OpenCue/compare/master...bcipriano:cuesubmit-no-shows-2?expand=1

Basically just collects errors during startup then raises them afterwards by using the widget's `showEvent`.

What do you think? Feel free to check out my branch and pull changes into yours if you'd like.

Reply:
Makes sense about the exception, didn't know why it would be useful before.

I modified 'messageBox' function in Widgets for similar cases. That function is not used too much and I made new argument optional, so I don't expect too high risk there.

I would prefer not to do this at all, hoped I could persuade it to work as in 'errorInJobData' where 'messageBox' is centered into parent widget properly. But as you mentioned, the issue is that app is not shown yet. I dived into exotic approaches how to solve it (not first one with similar issue on the internet), but failed miserably. I could probably persuade it with ugly hack with QTimer or something, but didn't want to over engineered it for hopefully very rare cases.

What do you think, should I try some hack workaround, or this would be enough for time being?

Reply:
Yeah, I tried a few things myself yesterday and came to the same conclusion. I think this is enough for the time being.

My only suggestion here would be -- rather than add `centerOnScreen` constructor argument, add a new method `centerOnScreen` instead. So the resulting calling code would look like:

```
msgBox = Widgets.messageBox(startupError, title="No Shows Exist", parent=self)
msgBox.show()
msgBox.centerOnScreen()
```

Reasoning here is, when I was playing around with it I found that I had to call `show()` before attempting to center it, otherwise the size/position would be off.

Reply:
I thought that '.sizeHint()' handled it, but you are right, it was only center-ish.

It wasn't constructor, as one would expect, but only function. I did small refactoring there to create class, name it according to existing names in 'Widget'.